### ISSUE TITLE
Using {{ $value }} doesnt return value

### ISSUE THREAD

**Issue by itsmjeu** [grafana/grafana#37393]
Using {{ $value }} on alert description returns:

lvl=eror msg="error in expanding template" logger=ngalert name=description value="Value is expected:  {{ $value }}" err="error parsing template __alert_Under_Replicated_Partitions: template: __alert_Under_Replicated_Partitions:1: u**ndefined variable \"$value\""**

While I was expecting to return the value 

To reproduce just add {{ $value }} on description of ngalert 

![image](https://user-images.githubusercontent.com/36747067/127635454-5b37177c-31e0-4ad8-9690-120b3bb5b381.png)

Environment:
- Grafana version: 8.0.6
- Data source type & version: Prometheus

**Comment by ymatsiuk** [comment]
Interesting bit is I see the value in preview of alert (`Info` field, even though it's truncated and not fully visible) but haven't found the way to access those fields.

**Comment by itsmjeu** [comment]
Exactly the value appears on preview but as soon you put {{ $value }} on the text description field or any other field, it stops processing everything else

**Comment by itsmjeu** [comment]
Any hope to access the values of the variables? 

{{ $value }} 

result in something like:

[ var='B' labels={clusterID=PKS-CLUSTER, instance=192.168.1.10:9100, namespace=pks-system, pod=node-exporter-sb6fw} value=19.99661364996668 ],
[ var='C' labels={clusterID=PKS-CLUSTER, instance=192.168.1.10:9100, namespace=pks-system, pod=node-exporter-sb6fw} value=0 ]

**Comment by nidhinkumar06** [comment]
For me also the same

**Comment by dmit2k** [comment]
> Any hope to access the values of the variables?
> 
> {{ $value }}
> 
> result in something like:
> 
> [ var='B' labels={clusterID=PKS-CLUSTER, instance=192.168.1.10:9100, namespace=pks-system, pod=node-exporter-sb6fw} value=19.99661364996668 ],
> [ var='C' labels={clusterID=PKS-CLUSTER, instance=192.168.1.10:9100, namespace=pks-system, pod=node-exporter-sb6fw} value=0 ]

According to https://grafana.com/docs/grafana/latest/alerting/unified-alerting/alerting-rules/create-grafana-managed-rule/ you should either use {{ $labels.YOURLABELNAME }} for label values or {{ $values.B.Labels.YOURLABELNAME }} for a label value from results of expression B or {{ $values.A.Value }} for the value of expression A itself, etc. 

Please note that this only works for the new Math / Reduce expression types, and does not work with Classic expressions (as of #37836)

**Comment by itsmjeu** [comment]
Yes that is the problem  {{ $values.A.Value }} doesn't work as I would expect to give me back the value of A

**Comment by itsmjeu** [comment]
ok I just used  {{ $values.A }}  was enough

**Comment by nidhinkumar06** [comment]
> ok I just used {{ $values.A }} was enough

Did you used in summary or description field.If so could you please show how it is done.I have tried the same in summary but i have received like below

```
summary = {{ $values.B }} {{ $values.B.Labels.device }}
```

![value](https://user-images.githubusercontent.com/10459220/130041570-a6134155-bc2a-4c1e-b4fb-3cb60ecded7d.png)

**Comment by itsmjeu** [comment]
It worked in both, I just tried out in the summary as well it was only on description my only problem now I not able to round it with {{ $values.B | printf "%.2f" }}

**Comment by itsmjeu** [comment]
It worked in both, I just tried out in the summary as well it was only on the description.
My only problem now I not able to round it with {{ $values.B | printf "%.2f" }} 

it throw me something like:


{map[%!f(string=gr):%!f(string=cc) %!f(string=to):%!f(string=cc)] %!f(*float64=0xc001a37830)}


I'm using version 8.1.1 by the way


It seems it is passing it as reference or something the results seems a memory position and not the value rounded to 2 decimal as expected

**Comment by nidhinkumar06** [comment]
@itsmjeu I was using 8.0.4 in which i tried with {{ $values.B }} and {{ $labels.cpu }} which doesn't returns the value.But when i upgrade it to 8.1.1 i was able to get the values.

**Comment by ymatsiuk** [comment]
@itsmjeu `{{ $values.B }}` is a map, so you probably want to refer to specific fields like so `{{ $values.B.Value }}` that one is float64 and I assume you can change formatting. As @dmit2k already pointed out, there is a section in [doc](https://grafana.com/docs/grafana/latest/alerting/unified-alerting/alerting-rules/create-grafana-managed-rule/#details) (scroll to "Template variables") now, so you can get more details there.

**Comment by itsmjeu** [comment]
Unfortunately the {{ $values.B.Value }} is not working here 

from:

Consumers lag is too high Group = {{$values.B.Labels.group }} Topic = {{$values.B.Labels.topic }} Value: {{ $values.B }} all the values = {{ $values }} test: {{ $values.B.Value | printf "%.2f" }}

output is: 

Consumers lag is too high Group = some-group Topic = some-topic Value: 36219.48168701443 all the values = map[B:36219.48168701443 c:1] 
with {{ $values.B.Value }}: %!f(*float64=0xc0049d0498)

I was expecting it to round the value

**Comment by gerobinson** [comment]
Hi @itsmjeu `$value` is not available in 8.0.6. It is an 8.1 feature which should explain the error message `undefined variable $value`.

**Comment by gerobinson** [comment]
I can see there is some confusion about the difference between `{{ $values.A }}` and `{{ $values.A.Value }}` and I don't think the documentation is clear enough here.

`{{ $values }}` is a [`map[string]templateCaptureValue`](https://github.com/grafana/grafana/blob/v8.1.0/pkg/services/ngalert/state/cache.go#L147). However `templateCaptureValue` implements the `fmt.Stringer` interface which means `{{ $values.A }}` is a string and not a float64 as the value can be absent. The raw value can be accessed with `{{ $values.A.Value }}` but this is a `*float64* as it can be `nil`, however I can see that its confusing some of the template functions which are printing the address of the pointer and not the value being referenced.

**Comment by gerobinson** [comment]
I can see there is some confusion about the difference between `{{ $values.A }}` and `{{ $values.A.Value }}` and I don't think the documentation is clear enough here.

`{{ $values }}` is a [`map[string]templateCaptureValue`](https://github.com/grafana/grafana/blob/v8.1.0/pkg/services/ngalert/state/cache.go#L147). However `templateCaptureValue` implements the `fmt.Stringer` interface which means `{{ $values.A }}` is a string and not a float64 as the value can be absent. The raw value can be accessed with `{{ $values.A.Value }}` but this is a `*float64* as it can be nil`, however I can see that its confusing some of the template functions which are printing the address of the pointer and not the value being referenced.

**Comment by gerobinson** [comment]
I can see there is some confusion about the difference between `{{ $values.A }}` and `{{ $values.A.Value }}` and I don't think the documentation is clear enough here.

`{{ $values }}` is a [`map[string]templateCaptureValue`](https://github.com/grafana/grafana/blob/v8.1.0/pkg/services/ngalert/state/cache.go#L147). However `templateCaptureValue` implements the `fmt.Stringer` interface which means `{{ $values.A }}` is a string and not a float64 as the value can be absent. The raw value can be accessed with `{{ $values.A.Value }}` but this is a `*float64` as it can be `nil`, however I can see that it's confusing some of the template functions which are printing the address of the pointer and not the value being referenced.

**Comment by gerobinson** [comment]
I can see there is some confusion about the difference between `{{ $values.A }}` and `{{ $values.A.Value }}` and I don't think the documentation is clear enough here.

`{{ $values }}` is a [`map[string]templateCaptureValue`](https://github.com/grafana/grafana/blob/v8.1.0/pkg/services/ngalert/state/cache.go#L147). However `templateCaptureValue` implements the `fmt.Stringer` interface which means `{{ $values.A }}` is a string and not a float64 as the value can be absent. The raw value can be accessed with `{{ $values.A.Value }}` but this is a `*float64` as it can be `nil`, however I can see that it's confusing some of the template functions which are printing the address held in the pointer and not the value being referenced.

**Comment by itsmjeu** [comment]
Any chance to have the {{ $values.A.Value }}  rounded some way?

**Comment by gerobinson** [comment]
@itsmjeu I don't think it's possible until we change it from `*float64` to `float64`? I have opened a PR to fix this above and hope to get it reviewed soon and in the next minor release.

**Comment by itsmjeu** [comment]
Ok super I will keep watch on it thanks! (#38766)

**Comment by pratikdas44** [comment]
> I can see there is some confusion about the difference between `{{ $values.A }}` and `{{ $values.A.Value }}` and I don't think the documentation is clear enough here.
> 
> `{{ $values }}` is a [`map[string]templateCaptureValue`](https://github.com/grafana/grafana/blob/v8.1.0/pkg/services/ngalert/state/cache.go#L147). However `templateCaptureValue` implements the `fmt.Stringer` interface which means `{{ $values.A }}` is a string and not a float64 as the value can be absent. The raw value can be accessed with `{{ $values.A.Value }}` but this is a `*float64` as it can be `nil`, however I can see that it's confusing some of the template functions which are printing the address held in the pointer and not the value being referenced.

Hi i am facing a similar issue can you look on this as well?
![image](https://user-images.githubusercontent.com/30981073/131691897-20b0759d-5dea-49e8-a530-05e686729386.png)

I tried with all possible options but none of them is printing.

My queries are like this - 
![image](https://user-images.githubusercontent.com/30981073/131692085-2cc938d3-825d-4526-8071-e0224a836058.png)
![image](https://user-images.githubusercontent.com/30981073/131692161-85679cbd-ac5b-4602-a311-5c99b3dd9950.png)

**Comment by nidhinkumar06** [comment]
@gerobinson does this issue is fixed?

**Comment by gerobinson** [comment]
This is fixed in 8.1.4!

**Comment by nidhinkumar06** [comment]
@gerobinson  I have installed the latest version 8.1.4 and tried to round off the value like below but it doesn't seem to be working. Kindly let me know whether it is the right fix.

![Screenshot from 2021-09-17 15-30-07](https://user-images.githubusercontent.com/10459220/133765087-bb0f6254-86b0-48d1-8a61-6ed66f0f8b57.png)


Alert

![Screenshot from 2021-09-17 15-32-18](https://user-images.githubusercontent.com/10459220/133765187-0600a8a8-7b76-4a08-aa26-f919b1cda8d2.png)

**Comment by gerobinson** [comment]
When expanding annotations and labels `$values.B` is a string as each value in `$values` is a struct that implements the `fmt.Stringer` interface. To use the floating point value with `printf` use `$values.B.Value`.

**Comment by nidhinkumar06** [comment]
@gerobinson  Like this we need to use printf ?

```
{{ $values.B.Value | printf "%.2f" }}
```

**Comment by pratikdas44** [comment]
@nidhinkumar06 are you using data source as Influxdb or something else.

**Comment by nidhinkumar06** [comment]
@pratikdas44 I am using InfluxDB as datasource

**Comment by ImmoWetzel** [comment]
Sorry to tell you that I wasnt able to use printf as proposed with Version v8.1.4 (6855cdff7c)

I used 
Available memory {{ $values.B.Value | printf "%.2f" }} GB is below 10% of total memory for instance {{ .Labels.short_instance}}
->
Available memory %!f(*float64=0xc0016a9fd8) GB is below 10% of total memory for instance host04

without printf i get

Available memory 7.728806100398056 GB is below 10% of total memory for instance host04

**Comment by ImmoWetzel** [comment]
@itsmjeu should I write a new ticket cos this one is closed

**Comment by gerobinson** [comment]
I forgot to select the `backport v8.1.x` label for this fix and as a result it didn't make it into the 8.1.4 release. This fix will instead be in the 8.2 final release. **I would not recommend upgrading to the 8.2 beta at this time as there is an issue where alerts created with ngalert on Grafana 8.1.x are deleted #39361**.

**Comment by mahmutefe** [comment]
Why Grafana is not providing a document regarding formats. I was struggling to print value for more than 3 hours. Still could not find a way to do it. I tried all types and still, it's not printing the correct data. 
I am using **v8.4.3** version.

`{{ $values.A.Value }}
{{ $values.A.Value | printf "%.2f" }}
{{ $values.A.value }}
{{ $values.A.calue | printf "%.2f" }}
{{ $values.A.labels.app }}`

responses are:

`<no value>
%!f(<nil>)
<no value>
%!f(<nil>)
<no value>`
